radix-transactions/src/manifest/static_resource_movements_visitor.rs

```diff
@@ -0,0 +1,2171 @@
+        };
+
+        // Getting all of the buckets and expressions in the arguments.
+        let (buckets, expressions) = {
+            let encoded = match manifest_encode(args) {
+                Ok(encoded) => encoded,
+                Err(error) => {
+                    return ControlFlow::Break(
+                        ManifestValidationError::ArgsEncodeError(error).into(),
+                    )
+                }
+            };
+            let mut traverser = ManifestTraverser::new(
+                &encoded,
+                ExpectedStart::PayloadPrefix(MANIFEST_SBOR_V1_PAYLOAD_PREFIX),
+                VecTraverserConfig {
+                    max_depth: MANIFEST_SBOR_V1_MAX_DEPTH,
+                    check_exact_end: true,
+                },
+            );
+            let mut buckets = index_set_new();
+            let mut expressions = index_set_new();
+            loop {
+                let event = traverser.next_event();
+                match event.event {
+                    TraversalEvent::TerminalValue(value) => match value {
+                        TerminalValueRef::Custom(ManifestCustomTerminalValueRef(
+                            ManifestCustomValue::Bucket(bucket),
+                        )) => {
+                            buckets.insert(bucket);
+                        }
+                        TerminalValueRef::Custom(ManifestCustomTerminalValueRef(
+                            ManifestCustomValue::Expression(expression),
+                        )) => {
+                            expressions.insert(expression);
+                        }
+                        _ => {}
+                    },
+                    TraversalEvent::ContainerStart(_)
+                    | TraversalEvent::ContainerEnd(_)
+                    | TraversalEvent::TerminalValueBatch(_) => {}
+                    TraversalEvent::DecodeError(error) => {
+                        return ControlFlow::Break(
+                            ManifestValidationError::ArgsDecodeError(error).into(),
+                        );
+                    }
+                    TraversalEvent::End => break,
+                }
+            }
+            (buckets, expressions)
+        };
+
+        // Resolving the buckets and the expressions into the bucket contents. We do not consume the
+        // buckets here, we just get a clone of the contents of the bucket. This is because there is
+        // a specific callback for consuming buckets.
+        let mut invocation_inputs = result_to_control_flow(
+            buckets
+                .into_iter()
+                .map(|bucket| {
+                    self.tracked_buckets
+                        .get(&bucket)
+                        .cloned()
+                        .map(InvocationIo::from)
+                        .ok_or(StaticResourceMovementsError::BucketDoesntExist(bucket))
+                })
+                .collect::<Result<Vec<_>, _>>(),
+        )?;
+        if expressions
```

Comment:
From a code organisation perspective, this method is very long. And whilst the code itself is very nice to read, I'd suggest breaking it up into separate methods would make it easier to jump around and reason about.

In fact, I'd suggest the `on_x` methods just call out to `self.handle_x(on_x)` and then you can break up the method in the confines of a normal impl block.

e.g. a method tree might look like:
* `handle_instruction_end`
    * `finalize_inbound_resources`
    * `resolve_outbound_resources`
        * `resolve_native_invocation_kind`
        * `resolve_native_inbound_resources`
            * `resolve_account_inbound_resources`
                * `inbound_fungible_amount`
                * `inbound_nonfungible_amount`
            * `resolve_access_controller_inbound_resources` 
                * `inbound_fungible_amount`
            ...

I take this to pretty extreme extents in the `StaticManifestInterpreter` but I think this makes it pretty readable and keeps methods short and singular in their focus. But then again, I really like Uncle Bob style code! 👍

Reply:
I just split some of the logic into separate methods to make it all smaller.